clients/git/client.go

```diff
@@ -0,0 +1,208 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package git defines helper functions for clients.RepoClient interface.
+package git
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	cp "github.com/otiai10/copy"
+
+	"github.com/ossf/scorecard/v4/clients"
+)
+
+const repoDir = "repo*"
+
+var (
+	errNilCommitFound = errors.New("nil commit found")
+	errEmptyQuery     = errors.New("query is empty")
+)
+
+type Client struct {
+	gitRepo        *git.Repository
+	worktree       *git.Worktree
+	listCommits    *sync.Once
+	tempDir        string
+	errListCommits error
+	commits        []clients.Commit
+	commitDepth    int
+}
+
+func (c *Client) InitRepo(uri, commitSHA string, commitDepth int) error {
+	// cleanup previous state, if any.
+	c.Close()
+	c.listCommits = new(sync.Once)
+	c.commits = nil
+
+	// init
+	c.commitDepth = commitDepth
+	tempDir, err := os.MkdirTemp("", repoDir)
+	if err != nil {
+		return fmt.Errorf("os.MkdirTemp: %w", err)
+	}
+
+	// git clone
+	const filePrefix = "file://"
+	if strings.HasPrefix(uri, filePrefix) {
+		if err := cp.Copy(strings.TrimPrefix(uri, filePrefix), tempDir); err != nil {
+			return fmt.Errorf("cp.Copy: %w", err)
+		}
+		c.gitRepo, err = git.PlainOpen(tempDir)
+		if err != nil {
+			return fmt.Errorf("git.PlainOpen: %w", err)
+		}
+	} else {
+		c.gitRepo, err = git.PlainClone(tempDir, false /*isBare*/, &git.CloneOptions{
+			URL:      uri,
+			Progress: os.Stdout,
+		})
+		if err != nil {
+			return fmt.Errorf("git.PlainClone: %w", err)
+		}
+	}
+	c.tempDir = tempDir
+	c.worktree, err = c.gitRepo.Worktree()
+	if err != nil {
+		return fmt.Errorf("git.Worktree: %w", err)
+	}
+
+	// git checkout
+	if commitSHA != clients.HeadSHA {
+		if err := c.worktree.Checkout(&git.CheckoutOptions{
+			Hash:  plumbing.NewHash(commitSHA),
+			Force: true, // throw away any unsaved changes.
+		}); err != nil {
+			return fmt.Errorf("git.Worktree: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) ListCommits() ([]clients.Commit, error) {
+	c.listCommits.Do(func() {
+		commitIter, err := c.gitRepo.Log(&git.LogOptions{
+			Order: git.LogOrderCommitterTime,
+		})
+		if err != nil {
+			c.errListCommits = fmt.Errorf("git.CommitObjects: %w", err)
+			return
+		}
+		for i := 0; i < c.commitDepth; i++ {
+			commit, err := commitIter.Next()
+			if err != nil && !errors.Is(err, io.EOF) {
+				c.errListCommits = fmt.Errorf("commitIter.Next: %w", err)
+				return
+			}
+			// No more commits.
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			if commit == nil {
+				// Not sure in what case a nil commit is returned. Fail explicitly.
+				c.errListCommits = fmt.Errorf("%w", errNilCommitFound)
+				return
+			}
+
+			c.commits = append(c.commits, clients.Commit{
```

Comment:
I prefer dynamic allocation here since there is a possibility that there are not enough commits for `commitDepth`. In which case `len([]commit) == commitDepth` with lot of empty slots. This then puts the burden on caller code to handle this.

Reply:
`len([]commit) == commitDepth` isn't true if you use explicity set the length to 0: `make([]commit, 0, commitDepth)`

In this case: `cap([]commit) == commitDepth`.

There is a chance to over allocate space, but the caller code won't need to handle empty slots.

Reply:
Do you mean something like this?

Reply:
Yeah, thanks